vispy/visuals/filters/mesh.py

```diff
@@ -0,0 +1,323 @@
+        self._light_color = Color(light_color)
+        self._update_data()
+
+    @property
+    def diffuse_color(self):
+        """The diffuse light color."""
+        return self._diffuse_color
+
+    @diffuse_color.setter
+    def diffuse_color(self, diffuse_color):
+        self._diffuse_color = Color(diffuse_color)
+        self._update_data()
+
+    @property
+    def specular_color(self):
+        """The specular light color."""
+        return self._specular_color
+
+    @specular_color.setter
+    def specular_color(self, specular_color):
+        self._specular_color = Color(specular_color)
+        self._update_data()
+
+    @property
+    def ambient_color(self):
+        """The ambient color."""
+        return self._ambient_color
+
+    @ambient_color.setter
+    def ambient_color(self, color):
+        self._ambient_color = Color(color)
+        self._update_data()
+
+    @property
+    def shininess(self):
+        """The shininess."""
+        return self._shininess
+
+    @shininess.setter
+    def shininess(self, shininess):
+        self._shininess = shininess
+        self._update_data()
+
+    def _update_data(self):
+        if not self._attached:
+            return
+        self.vcode['light_dir'] = self._light_dir
+        self.fcode['shininess'] = self._shininess
+        self.fcode['light_color'] = self._light_color.rgb
+        self.fcode['ambient_color'] = self._ambient_color.rgb
+        self.fcode['diffuse_color'] = self._diffuse_color.rgb
+        self.fcode['specular_color'] = self._specular_color.rgb
+        self.fcode['flat_shading'] = 1 if self._shading == 'flat' else 0
+        self.fcode['shading_enabled'] = 1 if self._shading is not None else 0
+        normals = self._visual().mesh_data.get_vertex_normals(indexed='faces')
+        self._normals.set_data(normals, convert=True)
+
+    def _attach(self, visual):
+        # vertex shader
+        vert_post = visual._get_hook('vert', 'post')
+        vert_post.add(self.vertex_program)
+        render2scene = visual.transforms.get_transform('render', 'scene')
+        visual2scene = visual.transforms.get_transform('visual', 'scene')
+        scene2doc = visual.transforms.get_transform('scene', 'document')
+        doc2scene = visual.transforms.get_transform('document', 'scene')
+        self.vcode['render2scene'] = render2scene
+        self.vcode['visual2scene'] = visual2scene
+        self.vcode['scene2doc'] = scene2doc
+        self.vcode['doc2scene'] = doc2scene
+
+        # fragment shader
+        frag_post = visual._get_hook('frag', 'post')
+        frag_post.add(self.fragment_program)
+
+        self._attached = True
+        self._visual = weakref.ref(visual)
+        self._update_data()
+
+    def _detach(self, visual):
+        visual._get_hook('vert', 'post').remove(self.vertex_program)
+        visual._get_hook('frag', 'post').remove(self.fragment_program)
+
+
+wireframe_vertex_template = """
+varying vec3 v_bc;
+
+void preprare_wireframe() {
```

Comment:
Typo here `prepare`, is this ever used?

Reply:
If `prepare_wireframe()` is named `main()` instead, it produces a linking error:
```
[...]
RuntimeError: Program linking error:
error: fragment shader varying v_base_color not written by vertex shader
.error: fragment shader varying v_eye_vec not written by vertex shader
.error: fragment shader varying v_light_vec not written by vertex shader
.error: fragment shader varying v_normal_vec not written by vertex shader
.error: fragment shader varying v_pos_scene not written by vertex shader
```
I assume the main function of the original shader gets overridden.

Is there any documentation of the shader parsing and hook system mechanism?

My observations are:
- The `prepare_wireframe()` is being called, otherwise the varying would not be initialised in the fragment shader; the shader would not produce the wireframe.
- Any other unique name seems OK, e.g. "pass_bc_coords()`.
- The first function defined in the hook is used as the entry point. For example, this works:
```
varying vec3 v_bc;

void preprare_wireframe() {
    v_bc = $bc;
}

void other_func () {
}
```
This does not:
```
varying vec3 v_bc;

void other_func () {
}

void preprare_wireframe() {
    v_bc = $bc;
}
```

Reply:
All I'm saying is that the function says `preprare` when it is trying to say `prepare` (note the extra 'r').

Reply:
Right. Thanks!